### PR TITLE
Fix MCMC getting stuck, and speed up matrix exponential.

### PR DESCRIPTION
**Pull request by bredelings**
This PR was in response to a request from @jembrown, where MCMC wasn't working with the WAG model.  The PR switches from the _eigenvalue_-based matrix-exponential to the _scaling-and-squaring_ matrix-exponential that is based on a Taylor expansion.  This reportedly solves the problem with the MCMC getting stuck.

The issue is (probably) that the matrix exponential can return bad values.  If the rows sum to more than 1, then this can lead to large likelihoods even when the likelihood should be quite small.  Then the MCMC gets stuck on a large branch length.  

Such bad values can happen when branch lengths are large, or (under the eigenvalue-based matrix-exponential) when some of the aa-frequencies are small.  

The patch does a number of things.  It
* rewrites the WAG model to use the `GenericFunction` framework.
* rewrites the LG model to use the `GenericFunction` framework.
* allows estimating equilibrium frequencies with `fnLG(pi)`
* simplifies the GenericFunction framework
* switches WAG and LG to use AbstractRateMatrix::exponentialByScalingAndSquaring( ).

It also
* speeds up operations on `TransitionProbabilityMatrix` by inlining basic operations.
* implements `operator*` for TransitionProbabilityMatrix.
* removes `AbstractRateMatrix::multiplyMatrices` and uses TransitionProbabilityMatrix::operator* instead.
* improves mathematical stability for exponentiateMatrixByScalingAndSquaring
     * computes `p_4 = p_2 * p_2` instead of `p * p_3`
     * computes the required number of squarings from the L1 norm of `Q*t`.
     * coerces the result of the Taylor approximation into stochastic matrix form (all positive entries, rows sum to one)
     * decreases round-off error by renormalizing rows to sum to one after each multiplication.

Arguably we should put the common code between the WAG and LG functions into a C++ `Func_Empirical(er, pi, name)` class.  But that could be done in a later patch, if we want more amino-acid matrices.

In the future we may want to drop `AbstractRateMatrix::exponentiateMatrixByScalingAndSquaring( )` in favor of Eigen or the Pade approximation used by the FreeK one.  But not yet, because neither of those is quite a drop-in replacement.

### CODE REVIEW

**Comment by jsigao** [comment]
Hi Ben @bredelings, these code changes look fine to me in general. I could look at them more carefully, but I'm a bit confused with one of the computational decision made here: reading from the code and your description in the PR, it seems that scaling and squaring will always be used to compute the P matrix under these two models (WAG+F and LG+F). 

I guess I can see the necessity of using scaling and squaring to ensure numerical stability when some elements of the stationary frequencies are very close to zero---which is likely to appear during the MCMC---but it seems that computing the matrix exponential through eigendecomposition is more desirable when the frequency vector is fixed and none of them is too small (e.g., the default values), especially considering that the numerical stability and accuracy of the scaling and squaring method may be affected by the branch length.

In addition---perhaps this is tangential to this PR---given that both WAG and LG (with or without +F) are time reversible, in principle we could construct a symmetric matrix S that is similar to Q, i.e., S = A * Q * A^{-1}. Q is the instantaneous matrix and Q = R * \Pi; i.e., Q equals to the product of the symmetric exchangeability rates matrix R and the diagonal matrix \Pi, where the diagonal entries are the stationary frequencies. Then let A = \Pi^{1/2}, and thus S = \Pi^{1/2} * R * \Pi^{1/2}, so that S is symmetric. Therefore we could eigendecompose S = V * \Delta * V^{-1} (where V^{-1} is the transpose of V given that S is symmetric); and then Q = \Pi^{-1/2} * S * \Pi^{1/2} = ( \Pi^{-1/2} * V) * \Delta *  (V^{-1} * \Pi^{1/2}). Finally, we could compute P(t) = (\Pi^{-1/2} * V) * diag(exp(\delta * t))  *  (V^{-1} * \Pi^{1/2}). As S is symmetric, this computation should be more stable and no complex eigenvalues or eigenvectors would be involved. 

Again, these are just some of my thoughts on one of the potential more long-term solutions that could improve the numerical performance of this issue. I would be fine with the current solution that uses the scaling and squaring algorithm if you would prefer that to be merged into the development right away.

**Comment by bredelings** [comment]
Hi @jsigao, thanks for your comments!

In terms of speed, you are perhaps right that it would be better to use an eigenvalue approach except when some of the aa-frequencies are very small.  The main reason I switched to scaling-and-squaring for all cases was that it was simpler to code, and it reportedly fixed the problem, and I don't have a test case.  Conditionally updating the eigensystem might be complicated.  (As an aside, I don't think anybody wants to use WAG or LG without estimating the amino-acid frequencies.)

Yeah, I know about the reversible-eigenvalue approach.  I used it in my own code until recently.  I never had any problems with it getting stuck anywhere, although that is probably because I renormalized the rows.  It does have problems with small pi[i], and also (if I remember correctly) does not do a good job if the matrix is sparse.  It is possible that the existing eigenvalue code in revbayes would mostly work if we just renormalize the rows -- but I need to get my hands on the testcase in order to check.

That is an interesting point you make about large branch lengths.  I guess you are thinking that for large branch lengths, the scaling-and-squaring code needs to do a lot of squarings, and that could create a lot of round-off error?  After the improvements I made, I am able to get reasonable results up to `fnWAG.getTransitionProbabilities(1e300)`.  However, going up to `1e320` leads to NaNs -- I have not checked where the NaNs are coming from.  

I can see two approaches.  But are fine with me:
* merge this for now, and work on speed/stability issues as a follow-up.
* wait to get a test-case from @jembrown, and see if the MCMC can be fixed in a better/simpler way.  Since this PR has accumulated a number of other fixes, I would probably split those pieces out and try to get them merged sooner.

Or perhaps you have another idea.

Thoughts?

-BenRI

P.S. For the eigenvalue approach, I've found that using expm1 for small values of t can be helpful:
[expm1.pdf](https://github.com/revbayes/revbayes/files/9830468/expm1.pdf)

**Comment by jsigao** [comment]
Hi @bredelings, thanks for your prompt reply!

> Yeah, I know about the reversible-eigenvalue approach. I used it in my own code until recently, and I think most non-RevBayes software uses that approach... P.S. For the eigenvalue approach, I've found that using expm1 for small values of t can be helpful:
[expm1.pdf](https://github.com/revbayes/revbayes/files/9830468/expm1.pdf)

This is great to know! The writeup about expm1 is very helpful. This reminds me about the uniformization approach (e.g., [Rodrigue et al., 2008](https://academic.oup.com/bioinformatics/article/24/1/56/205240); [Tataru and Hobolth, 2011](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-465)) where the matrix exponential can be computed via the multiplication of the uniformized matrix (which is implemented as one of matrix-exponential methods for the FreeK matrix, see [here](https://github.com/revbayes/revbayes/blob/f113987142f79603644632e68ca528c26672887d/src/core/datatypes/phylogenetics/ratematrix/RateMatrix_FreeK.cpp#L602)). 

Anyway, I think we agree that other improvements could be made to help this matrix exponential issue and it would be interesting to implement them later, but that should not be the reason of deferring the acceptance of this PR.

> That is an interesting point you make about large branch lengths... After the improvements I made, I am able to get reasonable results [1] up to fnWAG().getTransitionProbabilities(1e305). However, going up to 1e306 leads to NaNs -- presumably because 1e310 evaluates to Inf.

This is great to know as well! I guess I just meant in principle my impression was that the scaling and squaring methods could be sensitive to the scale of the norm of the matrix, and longer branch length may require more rounds of multiplication and more likely to lose accuracy, but it's great that you have demonstrated that it's actually very robust here.

> In terms of speed, you are perhaps right that it would be better to use an eigenvalue approach except when some of the aa-frequencies are very small. The main reason I switched to scaling-and-squaring for all cases was that it was simpler to code, and it reportedly fixed the problem, and I don't have a test case. Conditionally updating the eigensystem might be complicated. (As an aside, I don't think anybody wants to use WAG or LG without estimating the amino-acid frequencies.)

This may be my only reservation before approving this PR entirely. The original behavior (before this PR) of the WAG and LG matrices are identical to the other empirical matrices, where the eigensystem is computed once upon initialization, and will be updated upon a change of the frequency vector (i.e., it only needs to be computed once if the frequency vector is fixed instead of estimated). What I meant in my previous comment is that it seemed to me that---despite that most users will be estimating the stationary frequencies---we might still prefer to keep that default behavior and make the program switch to scaling and squaring when the stationary frequencies are estimated.

To step back, I don't feel very strongly about this and perhaps the preference of the people who use these empirical matrices are more important. However, my slight preference would be keeping the WAG and LG datatype classes (i.e., not removing RateMatrix_Wag and RateMatrix_Lg) to allow future possibilities (as some certain model-specific properties of the empirical-matrix model that could help the computation could be stored in each specific rate-matrix datatype class).

**Comment by bredelings** [comment]
> This may be my only reservation before approving this PR entirely. The original behavior (before this PR) of the WAG and LG matrices are identical to the other empirical matrices, where the eigensystem is computed once upon initialization, and will be updated upon a change of the frequency vector (i.e., it only needs to be computed once if the frequency vector is fixed instead of estimated).  What I meant in my previous comment is that it seemed to me that---despite that most users will be estimating the stationary frequencies---we might still prefer to keep that default behavior and make the program switch to scaling and squaring when the stationary frequencies are estimated.

Hmm... that would probably work.  Again, I don't think it will be very useful, but it probably would not be too complicated.  Basically, if the user calls `fnWAG()` then we use the eigensystem-based function, and if they call `fnWAG(pi)` then we use the scaling-and-squaring based function.

Note that, after talking with @jembrown, the plan is to eventually modify the other empirical matrices to estimate the frequency vector as well.

> To step back, I don't feel very strongly about this and perhaps the preference of the people who use these empirical matrices are more important. However, my slight preference would be keeping the WAG and LG datatype classes (i.e., not removing RateMatrix_Wag and RateMatrix_Lg) to allow future possibilities (as some certain model-specific properties of the empirical-matrix model that could help the computation could be stored in each specific rate-matrix datatype class).

I'm not sure if I understand this part.  If you want to cache computations like the eigensystem, then I can see that you might need a datatype class to store the eigensystem.  However, I don't see why you would need a _different_ class for WAG and LG.  You could make a single class (maybe RateMatrix_EmpiricalEigensystem) and share it between all the empirical matrix models that want to cache an eigensystem.  Am I missing something?

File: src/revlanguage/functions/phylogenetics/ratematrix/Func_wag.cpp [hunk 1]
```diff
@@ -72,7 +182,6 @@ Func_wag* Func_wag::clone( void ) const
 
 RevBayesCore::TypedFunction< RevBayesCore::RateGenerator >* Func_wag::createFunction( void ) const
 {
-    
     RevBayesCore::TypedDagNode<RevBayesCore::Simplex>* bf = NULL;
     const RevObject* obj = &this->args[0].getVariable()->getRevObject();
     if (obj == NULL) {
```
Comment (jsigao) [inline src/revlanguage/functions/phylogenetics/ratematrix/Func_wag.cpp:187]: It seems that obj will never be NULL as the default value of this argument is the empirical frequency vector?

**Review by jsigao — COMMENTED** [review]
>I'm not sure if I understand this part. If you want to cache computations like the eigensystem, then I can see that you might need a datatype class to store the eigensystem. However, I don't see why you would need two different classes for WAG and LG. You could make a single class (maybe RateMatrix_EmpiricalEigensystem) and share it between all the empirical matrix models that want to cache an eigensystem.

I guess I meant that in principle all the empirical-matrix models could inherit from the RateMatrix_Empirical class, and each specific empirical-matrix class (e.g., RateMatrix_Wag) could store the corresponding hard coded eigensystem (including the eigenvalues and the C_{ijk}s) of the fixed matrix (when the frequencies are not estimated). Then, as you mentioned, there could a check in the RateMatrix_Empirical where the scaling and squaring method will be used if the frequencies are estimated; otherwise when the Q matrix is fixed the hard coded eigensystem would be used to compute the matrix exponential.

Nevertheless, as you pointed out, in practice if most users will be estimating the frequencies then it might not be worthwhile to optimize for the case where the Q matrix is entirely fixed. I'm fine with always using the more robust scaling and squaring method for now.

I just went over the commits again and added a couple superficial comments. I'm ready to approve this PR once they are responded to.

File: src/revlanguage/functions/phylogenetics/ratematrix/Func_wag.cpp (hunk 1 above)
Comment (bredelings) [reply src/revlanguage/functions/phylogenetics/ratematrix/Func_wag.cpp]: This was old code.  I cleaned it up a bit, and it is smaller and simpler now.

File: src/core/datatypes/phylogenetics/ratematrix/AbstractRateMatrix.cpp [hunk 2]
```diff
@@ -543,7 +579,12 @@ void AbstractRateMatrix::exponentiateMatrixByScalingAndSquaring(double t,  Trans
     // efficiency and returned the same results with about 10^-9 accuracy. The scaling parameter could be
     // increased for better accuracy.
     // -- Will Freyman 11/27/16
-    size_t s = 8;
+    double norm = L1norm(*the_rate_matrix) * std::abs(t);
+
+    int s = 0;
+    std::frexp(norm/0.005, &s);
```
[the remark this answers is not in the document]
Comment (bredelings) [reply src/core/datatypes/phylogenetics/ratematrix/AbstractRateMatrix.cpp:590]: I added an explanation as a comment.  The comment explains the limitation of the values `0.005` and `8`.

[the remark this answers is not in the document]
Comment (bredelings) [reply src/core/datatypes/phylogenetics/ratematrix/AbstractRateMatrix.cpp]: I simplified the formula somewhat, and explained the rationale.

**Review by jsigao — APPROVED** [review]
Thanks @bredelings! This PR looks all good to me now, so I'm approving it.

**Comment by bredelings** [comment]
Thanks Jiansi!

**Pull request merged**